.github/workflows/main.yml

```diff
@@ -7,30 +7,112 @@ on:
     # run daily, this refreshes the cache
     - cron: '13 2 * * *'
 
+concurrency:  # On new push, cancel old workflows from the same PR, branch or tag:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
 jobs:
   python-test:
     name: Python tests
     runs-on: ubuntu-22.04
     strategy:
       fail-fast: false
       matrix:
-        docker-image:  # See https://hub.docker.com/_/python/tags for images
-          - python:2.7.18-alpine3.11
-          - python:3.11.4-alpine3.18
-    container: ${{ matrix.docker-image }}
+        python-version: ['2.7', '3.11']
```

Comment:
Since all Python files have already been migrated to Python 3, do we still need to continue using Python 2's unittest? If so, there might be some adjustments needed within the unittest.

For instance, Python 2 does not support Chinese characters, but for the sake of code coverage, we need to include Chinese characters in our code. In such cases, Python 2 unittest might raise 'Non-ASCII character' error, and we might need to write some subtle code to work around it.